arch/x86/crc32_fold_pclmulqdq_tpl.h

```diff
@@ -41,12 +40,12 @@ Z_INTERNAL void CRC32_FOLD(crc32_fold *crc, const uint8_t *src, size_t len, uint
     crc32_fold_load((__m128i *)crc->fold, &xmm_crc0, &xmm_crc1, &xmm_crc2, &xmm_crc3);
 
     if (len < 16) {
-#ifdef COPY
```

Comment:
Is this necessary I wonder. We (currently) don't pass it lengths shorter than 16, so why move the ifdef here?

Reply:
Experimentally put the whole if block behind #ifdef:
```C
#ifdef COPY
    if (len < 16) {
        if (len == 0)
            return;

        memcpy(partial_buf, src, len);
        xmm_crc_part = _mm_load_si128((const __m128i *)partial_buf);
        memcpy(dst, partial_buf, len);
        goto partial;
    }
#endif
```

```
crc32/pclmulqdq/1             5.06 ns         5.05 ns    138511467
crc32/pclmulqdq/8             18.7 ns         18.6 ns     37546557
crc32/pclmulqdq/12            27.8 ns         27.7 ns     25243641
crc32/pclmulqdq/16            24.2 ns         24.2 ns     28941186
crc32/pclmulqdq/32            25.4 ns         25.4 ns     27589066
crc32/pclmulqdq/64            28.4 ns         28.4 ns     24674583
crc32/pclmulqdq/512           49.5 ns         49.4 ns     14157868
crc32/pclmulqdq/4096           226 ns          226 ns      3100691
crc32/pclmulqdq/32768         1609 ns         1606 ns       435830
crc32/pclmulqdq/262144       13585 ns        13558 ns        51634
crc32/pclmulqdq/4194304     226585 ns       226110 ns         3097

   text    data     bss     dec     hex filename
 133140    1352       8  134500   20d64 libz-ng.so.2
```
Speeds up especially the shorter buffers.
@KungFuJesus Is there any reason not to do this? Would it even work with lengths shorter than 16?

We should also update the above assert and comment.

Reply:
From what I found testing that it failed the gtests for lengths <16. I _think_ it fails to work when called with an initial checksum because the logic to deal with it is happening after that first block (that xor_initial stuff).

Reply:
Then isn't it enough to have the assert trigger at below 16, and apply the change I suggested above?
If it can't do <16, and we don't even call it <16, then there is no need for a block handling <16, right?

Reply:
The block for handling input < 16 is specifically for the case of a folded checksum where we are rolling a x bytes at a time into the checksum.  It definitely needs to handle those cases.  This could could likely be modified to handle arbitrary lengths but that XOR_INITIAL logic that's used to deal with an initial checksum fed to the function (via the init_crc parameter) would need to be reworked.

Reply:
Ah, now it makes sense 👍 

I think it is very ok to use crc32_small for the len<16 case btw. The speed is now quite good and even for all lengths, and any gains would probably not be a lot for those shorter lengths anyway.